Move to building with the 9.0 SDK and move our ToolCurrent TFM to 9.0

jkoritzinsky opened the pull request:
Unblocks https://github.com/dotnet/runtime/pull/95904

jkoritzinsky committed:
Move to building with the 9.0 SDK and move our ToolCurrent TFM to 9.0

am11 commented:
Looks like `src/tools/illink/test/Trimming.Tests.Shared/PathUtilities.cs` can be simplified to:
```diff
        public static class PathUtilities
        {
-#if NET8_0
-               public const string TFMDirectoryName = "net8.0";
-#elif NET7_0
-               public const string TFMDirectoryName = "net7.0";
-#elif NET6_0
-               public const string TFMDirectoryName = "net6.0";
-#elif NET5_0
-               public const string TFMDirectoryName = "net5.0";
-#elif NETCOREAPP3_0
-               public const string TFMDirectoryName = "netcoreapp3.0";
-#elif NET471
+#if NET471
                public const string TFMDirectoryName = "net471";
 #else
-#error "Unknown TFM"
+               public static readonly string TFMDirectoryName = $"net{Environment.Version.Major}.{Environment.Version.Minor}";
 #endif
```
(dropping netcoreapp3.0)

ViktorHofer committed:
Update CompatibilitySuppressions.xml

ViktorHofer committed:
Update PathUtilities.cs

ViktorHofer commented:
@sbomer @vitek-karas can you please take a look at the linker test failures?

ViktorHofer commented:
@mikem8361 can you please take a look at the failing `CustomAttributeDelete` test in System.Runtime.Loader?

mikem8361 commented:
I haven't work on ENC/hot reload/loader in 2 years. I assume some .NET 9 loader change broke this test. @mikelle-rogers might have an idea of what broken.

ViktorHofer committed:
Update TestProjects.targets

akoeplinger committed:
Merge branch 'main' into 9.0-sdk

ericstj commented:
I see @akoeplinger is updating https://github.com/dotnet/hotreload-utils/pull/348.  Maybe he's already onto what fix is needed here?

My best guess was that the hot reload diff somehow broke custom attribute type lookup.  Tried to debug, but these tests don't like that
<img width="451" alt="image" src="https://github.com/dotnet/runtime/assets/8918108/03ffb486-a144-49f1-9afd-3db6294100d6">

If @akoeplinger's fix to the hot reload tools doesn't resolve this, I recommend we just file an issue for the hot-reload owners to fix.

trylek commented:
I'm afraid I cannot add much more to what Eric has already discovered. Binary diffing of the ```dmeta``` file shows a "larger" diff for ```CustomAttributeDelete.dll.1.dmeta``` between the runtime repo without and with this change compared to the other dmeta files like ```CustomAttributeUpdate.dll.1.dmeta``` leading me to suspect that the tool used to generate the metadata delta files has a new bug in encoding the custom attribute deletion / replacement (the offending case is supposed to delete two attributes and add a new one) but I have no idea where its source code sits and / or how to dump the contents of the dmeta file that would let me investigate this further.

akoeplinger committed:
Merge branch 'main' into 9.0-sdk

akoeplinger committed:
Update hotreload-utils dependencies

akoeplinger committed:
Bump sdk version

akoeplinger commented:
Updating to latest hotreload-utils and newer SDK fixed this locally, let's see what CI says.

akoeplinger committed:
Update SourceBuildPrebuiltBaseline.xml

ViktorHofer committed:
Update ILLink.RoslynAnalyzer.Tests.csproj

ViktorHofer committed:
Update Mono.Linker.Tests.csproj

ViktorHofer committed:
Update CecilVersionCheck.cs

ViktorHofer committed:
Update PathUtilities.cs

ViktorHofer commented:
I simplified the path expectations logic in the linker tests but had to push individual commits (and manually) as my codespace didn't have permissions to push to this fork.

ViktorHofer committed:
Update PathUtilities.cs

ViktorHofer commented:
@radical please take a look at all the wasm/wasi failures

akoeplinger committed:
Merge branch 'main' into 9.0-sdk

sbomer committed:
Fix ILLink analyzer tests

am11 commented on src/tools/illink/test/ILLink.RoslynAnalyzer.Tests/TestCaseUtils.cs:
```diff
@@ -7,6 +7,7 @@
 using System.Diagnostics;
 using System.IO;
 using System.Linq;
+using System.Reflection;
```
```suggestion
```

am11 commented on src/tools/illink/test/ILLink.RoslynAnalyzer.Tests/TestCaseUtils.cs:
```diff
@@ -24,11 +25,15 @@ public abstract class TestCaseUtils
 		private static readonly string MonoLinkerTestsCases = "Mono.Linker.Tests.Cases";
 
+		private static string targetFramework = (string)AppContext.GetData("ILLink.RoslynAnalyzer.Tests.TargetFramework")!;
+
+		private static string frameworkVersion = typeof(object).Assembly.GetCustomAttribute<AssemblyInformationalVersionAttribute>()!.InformationalVersion.Split('+')[0];
```
```suggestion
```

am11 commented on src/tools/illink/test/ILLink.RoslynAnalyzer.Tests/TestCaseUtils.cs:
```diff
@@ -24,11 +25,15 @@ public abstract class TestCaseUtils
-				new PackageIdentity ("Microsoft.NETCore.App.Ref", "8.0.0"),
-				Path.Combine ("ref", "net8.0"))
+				targetFramework,
+				new PackageIdentity ("Microsoft.NETCore.App.Ref", frameworkVersion),
```
```suggestion
				new PackageIdentity ("Microsoft.NETCore.App.Ref", $"net{Environment.Version.Major}.{Environment.Version.Minor}"),
```

ViktorHofer replied:
This one isn't right. This should be the package version to restore.

ViktorHofer replied:
Fixed with the latest commit.

ViktorHofer committed:
Update src/tools/illink/test/ILLink.RoslynAnalyzer.Tests/TestCaseUtils.cs

Co-authored-by: Adeel Mujahid <3840695+am11@users.noreply.github.com>

ViktorHofer committed:
Update src/tools/illink/test/ILLink.RoslynAnalyzer.Tests/TestCaseUtils.cs

Co-authored-by: Adeel Mujahid <3840695+am11@users.noreply.github.com>

ViktorHofer committed:
Update src/tools/illink/test/ILLink.RoslynAnalyzer.Tests/TestCaseUtils.cs

Co-authored-by: Adeel Mujahid <3840695+am11@users.noreply.github.com>

ViktorHofer committed:
Remove NuGet restore for reference assemblies

ViktorHofer committed:
Fix ILCompiler tests

am11 commented:
@akoeplinger any idea about the iOS xharness failures?
```
[07:14:22] dbug: Looking for available ios-device devices. Storing logs into list-ios-device-20231229_071422.log
[07:14:25] crit: System.NullReferenceException: Object reference not set to an instance of an object.
                    at Microsoft.DotNet.XHarness.iOS.Shared.Hardware.Device.get_Architecture() in /_/src/Microsoft.DotNet.XHarness.iOS.Shared/Hardware/Device.cs:line 100
                    at Microsoft.DotNet.XHarness.iOS.Shared.Hardware.Device.get_Supports64Bit() in /_/src/Microsoft.DotNet.XHarness.iOS.Shared/Hardware/Device.cs:line 82
                    at Microsoft.DotNet.XHarness.iOS.Shared.Hardware.HardwareDeviceLoader.<>c.<get_Connected64BitIOS>b__7_0(IHardwareDevice x) in /_/src/Microsoft.DotNet.XHarness.iOS.Shared/Hardware/HardwareDeviceLoader.cs:line 48
                    at System.Linq.Enumerable.WhereEnumerableIterator`1.MoveNext()
                    at System.Linq.Enumerable.TryGetFirst[TSource](IEnumerable`1 source, Func`2 predicate, Boolean& found)
                    at System.Linq.Enumerable.FirstOrDefault[TSource](IEnumerable`1 source, Func`2 predicate)
                    at Microsoft.DotNet.XHarness.Apple.DeviceFinder.FindDevice(TestTargetOs target, String deviceName, ILog log, Boolean includeWirelessDevices, Boolean pairedDevicesOnly, CancellationToken cancellationToken) in /_/src/Microsoft.DotNet.XHarness.Apple/DeviceFinder.cs:line 87
                    at Microsoft.DotNet.XHarness.Apple.BaseOrchestrator.OrchestrateOperationInternal(TestTargetOs target, String deviceName, Boolean includeWirelessDevices, Boolean resetSimulator, Boolean enableLldb, GetAppBundleInfoFunc getAppBundle, ExecuteMacCatalystAppFunc executeMacCatalystApp, ExecuteAppFunc executeApp, CancellationToken cancellationToken) in /_/src/Microsoft.DotNet.XHarness.Apple/Orchestration/BaseOrchestrator.cs:line 215
                    at Microsoft.DotNet.XHarness.Apple.BaseOrchestrator.OrchestrateOperation(TestTargetOs target, String deviceName, Boolean includeWirelessDevices, Boolean resetSimulator, Boolean enableLldb, GetAppBundleInfoFunc getAppBundle, ExecuteMacCatalystAppFunc executeMacCatalystApp, ExecuteAppFunc executeApp, CancellationToken cancellationToken) in /_/src/Microsoft.DotNet.XHarness.Apple/Orchestration/BaseOrchestrator.cs:line 96
                    at Microsoft.DotNet.XHarness.Apple.RunOrchestrator.OrchestrateRun(GetAppBundleInfoFunc getAppBundleInfo, TestTargetOs target, String deviceName, TimeSpan timeout, TimeSpan launchTimeout, Int32 expectedExitCode, Boolean includeWirelessDevices, Boolean resetSimulator, Boolean enableLldb, Boolean signalAppEnd, Boolean waitForExit, IReadOnlyCollection`1 environmentalVariables, IEnumerable`1 passthroughArguments, CancellationToken cancellationToken) in /_/src/Microsoft.DotNet.XHarness.Apple/Orchestration/RunOrchestrator.cs:line 221
                    at Microsoft.DotNet.XHarness.CLI.Commands.Apple.AppleAppCommand`1.Invoke(ILogger logger) in /_/src/Microsoft.DotNet.XHarness.CLI/Commands/Apple/AppleAppCommand.cs:line 68
                    at Microsoft.DotNet.XHarness.CLI.Commands.XHarnessCommand`1.Invoke(IEnumerable`1 arguments) in /_/src/Microsoft.DotNet.XHarness.CLI/Commands/XHarnessCommand.cs:line 145
XHarness exit code: 71 (GENERAL_FAILURE)
```
most of the recent changes in dotnet/xharness repo are related to package updates.

ViktorHofer commented:
Alexander is currently on holiday. I'm not sure who else can take a look. Maybe @radical is already back and can take a look? cc @lewing

radical committed:
[wasi] Always add wasi-wasm rid to KnownRuntimePack for local builds in targetingpacks.targets

radical committed:
Merge remote-tracking branch 'origin/main' into 9.0-sdk

radical commented:
I'm working on fixes for the wasm failures.
1. wasi - smoke-aot - fixed
2. wasi/browser - WBT - WIP

radical committed:
wip

radical committed:
disable unrelated builds

akoeplinger commented:
@am11 that failure is a machine specific infrastructure issue, filed https://github.com/dotnet/dnceng/issues/1749 for that and https://github.com/dotnet/xharness/issues/1126 so that xharness doesn't fail in this case.

ViktorHofer commented:
@radical do you have an update on the remaining failures? We really need to get this in asap as arcade code flow is completely blocked right now. Thank you for taking a look.

radical committed:
Merge remote-tracking branch 'origin/main' into 9.0-sdk

radical commented:
Is this PR needed for the arcade bump or the reverse? The arcade PR has some similar failures for wasm. I'm trying to isolate the failures. I don't yet understand the reason for the new failures, like for wasm.build.tests we have assemblies being published from the wasm runtime pack, and thus they fail to run on desktop dotnet (like Process is not supported).

ViktorHofer commented:
> Is this PR needed for the arcade bump or the reverse?

This PR needs to go in before the "Update dependencies form arcade" PR. Ignore the Arcade dependency update PR for now.

radical committed:
split ci

radical committed:
Revert "split ci"

This reverts commit ebbc65763af9f8f1b01489afa520728db1db7e8b.

radical committed:
Revert "disable unrelated builds"

This reverts commit 01cf82aa8e3ff0e8054a03bfcfb66747927038db.

radical committed:
Revert "wip"

This reverts commit 275ead5e810e68b9e767e7cde636a4df768286bc.

radical committed:
Merge branch 'main' into 9.0-sdk

radical committed:
Merge remote-tracking branch 'origin/main' into 9.0-sdk

radical committed:
[wasm] Extract workload testing bits from tests.*.targets into separate files, so they can be used outside wasm test projects

radical committed:
[wasm] Wasm.Build.Tests - build for host os/arch instead of wasm

The tests themselves are built to run on coreclr, but each test would
build+run tests on wasm.

The CI builds are structured such that WBT gets built with
`TargetOS=browser` and `TargetArchitecture=wasm`, which would cause the
core assemblies to be bundled from the wasm runtime pack, resulting in
failures like:

```
      System.AggregateException : One or more errors occurred. (System.Diagnostics.Process is not supported on this platform.)
      ---- System.PlatformNotSupportedException : System.Diagnostics.Process is not supported on this platform.
      Stack Trace:
           at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
           at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)
           at System.Threading.Tasks.Task`1.get_Result()
```

To make sure that WBT is built for the host os/arch we clear the two
properties.

radical committed:
[wasm] fix local builds - makefile

radical committed:
[wasm] Run WBT with dotnet 8 (global.json)

.. instead of dotnet 9, with which it seems to segfault.

radical committed:
[wasm] fix local debugger tests run

radical committed:
[wasm] Browser.targets: track tfm change for tasks

radical committed:
[wasm] fix wbt build

radical commented:
/azp run runtime-wasm,runtime-extra-platforms

radical committed:
[wasm] WBT: don't use publishTrimmed=true

radical commented:
/azp run runtime-wasm

ViktorHofer commented on eng/targetingpacks.targets:
```diff
@@ -188,4 +188,13 @@
     </ItemGroup>
   </Target>
+
+  <Target Name="UpdateRuntimePackToAddWasiRID" BeforeTargets="ProcessFrameworkReferences">
+    <!-- Always add wasi-wasm since it isn't added by sdk at all -->
+    <ItemGroup>
+      <KnownRuntimePack Update="$(LocalFrameworkOverrideName)"
+        RuntimePackRuntimeIdentifiers="%(RuntimePackRuntimeIdentifiers);wasi-wasm"
+        Condition="'$(UseLocalTargetingRuntimePack)' == 'true' and '%(RuntimePackLabels)' == 'Mono' and $([MSBuild]::VersionGreaterThanOrEquals('$(TargetFrameworkVersion)', '8.0'))" />
+    </ItemGroup>
+  </Target>
```
Are you sure this is necessary?

wasi-wasm is already added to the live runtime pack: https://github.com/dotnet/runtime/blob/a54adf7f51b32d283bdd6f11767c065ed8c9a1e4/eng/targetingpacks.targets#L56C102-L56C102

radical replied:
IIUC, that is added only if the sdk didn't already have an entry for 9.0 . Because we are now using a sdk that has it, that line won't add anything. And the sdk does not itself add `wasi-wasm`, so we need to always add it unconditionally.

ViktorHofer replied:
Oh yes that makes sense. We should move this out of the target and make it use the same Update syntax as https://github.com/dotnet/runtime/blob/65fbeb45ab70ca62e5d161bdddf0bf9022f12543/eng/targetingpacks.targets#L51. I think it should be placed in the same ItemGroup just with a different condition.

`<KnownRuntimePack Update="@(KnownRuntimePack->WithMetadataValue('Identity', '$(LocalFrameworkOverrideName)')->WithMetadataValue('RuntimeBackLabels', 'Mono')` and remove the wasi-wasm entry from above so that we only set it once.

radical replied:
Outside the target I can do:
```xml
    <KnownRuntimePack Update="@(KnownRuntimePack->WithMetadataValue('Identity', '$(LocalFrameworkOverrideName)')->WithMetadataValue('RuntimePackLabels', 'Mono'))"
                      RuntimePackRuntimeIdentifiers="%(RuntimePackRuntimeIdentifiers);wasi-wasm"
                      Condition="'$(UseLocalTargetingRuntimePack)' == 'true'" />
```
.. but this should be done only for the runtime packs tfm>=8.0 . I can't use batching here, so I'm not sure how to restrict this with a condition `$([MSBuild]::VersionGreaterThanOrEquals('....', '8.0'))` to check against `KnownRuntimePack.TargetFramework`. Any suggestions?

ViktorHofer replied:
Do you need both 8.0 and 9.0? I thought we are already consistently targeting net9.0 everywhere for our tests. You could have two of these update entries, one for NetCoreAppCurrent and one for net8.0.

Asking as `UseLocalTargetingRuntimePack` suggests that we use the live runtime and targeting pack which is `net9.0`.

radical replied:
Good point. Only `net9.0` is required.

ViktorHofer commented on src/libraries/Directory.Build.props:
```diff
@@ -94,15 +94,14 @@
     <MonoEnvOptions Condition="'$(MonoEnvOptions)' == '' and '$(TargetsMobile)' != 'true' and '$(MonoForceInterpreter)' == 'true'">--interpreter</MonoEnvOptions>
   </PropertyGroup>
 
-  <PropertyGroup Condition="'$(TargetsMobile)' == 'true'">
```
This condition was there so that we don't set any of these properties unnecessarily when not testing workloads.

radical replied:
This was fine earlier when we were building Wasm/Wasi.Build.Tests with targetOS=wasi/browser, but now we are building them for the host os, and thus these properties would need to be available for `TargetsMobile!=true`.

The Was*Build.Tests run on coreclr, and each test then builds and runs test projects for wasm.

ViktorHofer replied:
Is there another feature flag that we can condition on here?

radical replied:
hmm not a feature flag really, but we could have a property to indicate that we are building a workload-testing dependent project, and use that here.

ViktorHofer replied:
I think that would be good.

radical commented:
One of the changes I had to make was to run the tests on helix with the default `dotnet` installed (version set by arcade?), instead of the latest (nearly daily) sdk because xunit segfault'ed when running with that.

radical commented:
I will debug on a test PR, and then push the fixes here.

radical committed:
Set workload testing specific properties in workloads-testing.targets

radical committed:
[wasm] Fix checking for icudt files against the runtime pack

The files in the appbundle were being compared against the file copied
to the bindir for Wasm?Build.Tests, like
`artifacts/bin/Wasm.Build.Tests/Release/net8.0/browser-wasm`.

But since WBT is not being built to target wasm anymore, these files are
never copied to the bindir. Instead, construct the path to runtime pack
and check against that. This is the correct behavior.

radical committed:
cleanup

radical committed:
WBT: Update TargetFrameworkForTasks=net9.0

radical committed:
WBT: Set chromedriver path

radical committed:
eng/targetingpack.targets: Add wasi-wasm only for the latest tfm

addresses review feedback from @ viktorhofer .

radical commented:
/azp run runtime-wasm

radical committed:
[wasm] WBT ProcessExtensions: Use Process.Kill(true)

.. instead of an old custom implementation.

radical committed:
[wasm] WBT: don't try to read ExitCode after process.Kill

`Process.Kill` returns immediately, and WaitOnExit needs to be used to
wait for the actual exit.

ViktorHofer commented on eng/targetingpacks.targets:
```diff
@@ -53,9 +53,13 @@
                       Condition="'$(UseLocalTargetingRuntimePack)' == 'true' and ('@(KnownRuntimePack)' == '' or @(KnownRuntimePack->WithMetadataValue('Identity', 'Microsoft.NETCore.App')->WithMetadataValue('RuntimePackLabels', 'Mono')->WithMetadataValue('TargetFramework', '$(NetCoreAppCurrent)')) == '')" />
+    <!-- always add wasi-wasm as it is never added by the sdk -->
+    <KnownRuntimePack Update="@(KnownRuntimePack->WithMetadataValue('Identity', '$(LocalFrameworkOverrideName)')->WithMetadataValue('RuntimePackLabels', 'Mono')->WithMetadataValue('TargetFramework', '$(NetCoreAppCurrent)'))"
+                      RuntimePackRuntimeIdentifiers="%(RuntimePackRuntimeIdentifiers);wasi-wasm"
```
Should this RID addition be up-streamed into dotnet/sdk?

radical replied:
wasi-wasm is only supported with the workload right now, so we add that as part of the workload. And adding here for in-tree builds. So, upstreaming it is not needed right now.

radical commented:
I have been trying to debug the last remaining failures on WBT, and it seems to be down to `dotnet new wasmbrowser` - the process never seems to end.

```
...
      [5734] [] [2024-01-12 03:34:50.792] [Debug] [Microsoft.TemplateEngine.Edge.Template.TemplateCreator] => [Execute] => [Template content generation]: Template content generation finished, took 71 ms
      [5734] [] The template "WebAssembly Browser App" was created successfully.
      [5734] []
      [5734] [] [2024-01-12 03:34:50.795] [Debug] [Template Engine] => [Execute]: Execute finished, took 234 ms--------------------
```

.. and we wait for 3mins but the process never seems to exit.

radical commented:
I can reproduce this only on CI. And it does not happen for the first few times/tests.

radical committed:
Merge branch 'main' into 9.0-sdk

ViktorHofer committed:
Update CompatibilitySuppressions.xml

radical committed:
[wasm] WBT: disable failing tests

Issue: https://github.com/dotnet/runtime/issues/97054

radical committed:
fix build

radical committed:
fix build

radical committed:
disable some missed tests

radical commented:
/azp run runtime-wasm

ViktorHofer commented:
- Two runtime-wasm legs timed out, even after retrying. This looks related to an AzDO machine pool issue (logs mention the machines got purged).
- An issue already exists for the libraries test failure.

I don't think that any of those issues block merging the PR. Given that we need to get this in ASAP for being able to ship .NET 9 Preview 1, I'm merging this in now.
